### DISCUSSION TITLE
Generate .exe python

### DISCUSSION THREAD

**Discussion by dmorafoguet** [SAP/PyRFC#374] [Q&A]
Please excuse my English,
I have a python app that works fine from vscode, I can connect to SAP and retrieve information.

When I generate an .exe with pyinstaller and run it, I get the following error: (any suggestions? Thanks!!)

During handling of the above exception, another exception occurred: Traceback (most recent call last): File "Gen_DBT_BRZ_SLV.py", line 1, in <module> File "PyInstaller\loader\pyimod02_importers.py", line 378, in exec_module File "packages\conexionSAP.py", line 1, in <module > File "PyInstaller\loader\pyimod02_importers.py", line 378, in exec_module File "pyrfc\__init__.py", line 11, in <module> File "importlib\metadata\__init__.py", line 1009, in version File "importlib\metadata\__init__.py", line 982, in distribution File "importlib\metadata\__init__.py", line 565, in from_name importlib.metadata.PackageNotFoundError: No package metadata was found for pyrfc

.SPEC

# -*- mode: python ; coding: utf-8 -*-

from PyInstaller.utils.hooks import collect_submodules
from PyInstaller.building.api import PYZ, EXE

a = Analysis(
    ['C:\\Automatizar_Mapeo_Campos\\Gen_DBT_BRZ_SLV.py'],
    pathex=['C:\\Automatizar_Mapeo_Campos\\packages'],
    binaries=[],
    datas=[
        ('C:\\Automatizar_Mapeo_Campos\\ficheros_input', 'ficheros_input'),
        ('C:\\Automatizar_Mapeo_Campos\\ficheros_output', 'ficheros_output')
    ],
    hiddenimports=['pyrfc', 'pyodbc'],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    noarchive=False,
    optimize=0,
)

pyz = PYZ(a.pure, a.zipped_data, cipher=None)

exe = EXE(
    pyz,
    a.scripts,
    a.binaries,
    a.datas,
    [],
    name='Genbrzslv',
    debug=True,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    upx_exclude=[],
    runtime_tmpdir=None,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)

**Comment by aizafaraz** [comment]
Hi,

I understand you're having trouble running your Python app as an .exe generated by PyInstaller, even though it works fine in VSCode. The error you’re encountering seems to be related to the pyrfc package not being included properly in the .exe.

Here are a few steps you can try to fix the issue:

**1. Include Hidden Imports:**
PyInstaller might not automatically detect all dependencies. You can manually specify hidden imports like this:


`pyinstaller --hidden-import=pyrfc --hidden-import=importlib.metadata Gen_DBT_BRZ_SLV.py
`

**2. Update Your .spec File:**
If you're using a **`.spec`** file, ensure the **`pyrfc`** package is included correctly. You can add it explicitly in the **`data`** section:

```
from PyInstaller.utils.hooks import collect_data_files
pyrfc_data_files = collect_data_files('pyrfc')

a = Analysis(
    ['C:\\Automatizar_Mapeo_Campos\\Gen_DBT_BRZ_SLV.py'],
    pathex=['C:\\Automatizar_Mapeo_Campos\\packages'],
    binaries=[],
    datas=[
        ('C:\\Automatizar_Mapeo_Campos\\ficheros_input', 'ficheros_input'),
        ('C:\\Automatizar_Mapeo_Campos\\ficheros_output', 'ficheros_output')
    ] + pyrfc_data_files,
    hiddenimports=['pyrfc', 'pyodbc', 'importlib.metadata'],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    noarchive=False,
    optimize=0,
)
```

**3. Check Virtual Environment:**
Make sure you're packaging your application from the same virtual environment where pyrfc is installed.

**4. Rebuild and Test:**
After making these changes, rebuild the .exe using the updated **`.spec`** file:


`pyinstaller your_spec_file.spec`

I hope this helps resolve the issue!

**Comment by ohnogaurav** [comment]
The error occurs because `pyrfc` metadata isn't included in the PyInstaller build. To fix it:

1. Add `pyrfc` to hidden imports in your `.spec` file:

    ```python
    hiddenimports=['pyrfc', 'pyodbc', 'importlib.metadata']
    ```

2. Create a `hook-pyrfc.py` file to include `pyrfc` metadata:

    ```python
    from PyInstaller.utils.hooks import copy_metadata
    datas = copy_metadata('pyrfc')
    ```

3. Add the hook path in your `.spec` file:

    ```python
    hookspath=['.']
    ```

4. Rebuild the `.exe` with:

    ```bash
    pyinstaller Gen_DBT_BRZ_SLV.spec
    ```
